src/metrics/include/metrics/v2/event.hpp

```diff
@@ -0,0 +1,282 @@
+#include <grpcpp/create_channel.h>
+#include <grpcpp/security/credentials.h>
+
+#include <metrics/operation.hpp>
+#include <poplarlib/collector.grpc.pb.h>
+
+/**
+ * @namespace genny::metrics::internals::v2 this namespace is private and only intended to be used
+ * by genny's own internals. No types from the genny::metrics::v2 namespace should ever be typed
+ * directly into the implementation of an actor.
+ */
+
+namespace genny::metrics {
+
+template <typename Clocksource>
+class OperationEventT;
+
+namespace internals::v2 {
+
+class PoplarRequestError : public std::runtime_error {
+public:
+    using std::runtime_error::runtime_error;
+};
+
+
+/**
+ * Manages the collector stub and its interactions.
+ */
+class CollectorStubInterface {
+public:
+    CollectorStubInterface() {
+        setStub();
+    }
+
+    poplar::PoplarEventCollector::StubInterface* operator->() {
+        return _stub.get();
+    }
+
+private:
+    // We should only have one channel across all threads.
+    static std::unique_ptr<poplar::PoplarEventCollector::StubInterface> _stub;
+
+    auto createStub() {
+        grpc::ChannelArguments args;
+        // The BDP estimator overwhelms the server with pings on a heavy workload.
+        args.SetInt(GRPC_ARG_HTTP2_BDP_PROBE, 0);
+        auto channel =
+            grpc::CreateCustomChannel("localhost:2288", grpc::InsecureChannelCredentials(), args);
+        _stub = poplar::PoplarEventCollector::NewStub(channel);
+        return true;
+    }
+
+    void setStub() {
+        static bool stub_created = createStub();
+        return;
+    }
+};
+
+
+/**
+ * Manages the stream of poplar EventMetrics.
+ */
+class StreamInterfaceImpl {
+public:
+    StreamInterfaceImpl()
+        : _options{}, _response{}, _context{}, _stream{_stub->StreamEvents(&_context, &_response)} {
+        _options.set_no_compression().set_buffer_hint();
+    }
+
+    void write(const poplar::EventMetrics& event) {
+        auto success = _stream->Write(event, _options);
+
+        if (!success) {
+            BOOST_THROW_EXCEPTION(PoplarRequestError("Failed to write to stream."));
+        }
+    }
+
+    ~StreamInterfaceImpl() {
+        if (!_stream) {
+            BOOST_LOG_TRIVIAL(error) << "Tried to close grpc stream, but none existed.";
+            return;
+        }
+        if (!_stream->WritesDone()) {
+            BOOST_LOG_TRIVIAL(warning) << "Closing grpc stream, but not all writes completed.";
+        }
+        auto status = _stream->Finish();
+        if (!status.ok()) {
+            BOOST_LOG_TRIVIAL(error)
+                << "Problem closing grpc stream: " << _context.debug_error_string();
+        }
+    }
+
+private:
+    CollectorStubInterface _stub;
+    grpc::WriteOptions _options;
+    poplar::PoplarResponse _response;
+    grpc::ClientContext _context;
+    std::unique_ptr<grpc::ClientWriterInterface<poplar::EventMetrics>> _stream;
+};
+
+/**
+ * Very basic mock for tests.
+ */
+class MockStreamInterface {
```

Comment:
is it feasible to put this impl into the tests rather than letting test code (albeit simple test code) live in production?